Move mimetypes.list.php to config/mimetypemapping.dist.json

rullzer opened the pull request:
This allows users to add new mimetypemappings (extention -> mimetype)
themself. And not have to wait until a new release for updated
mimetypes.

Fixes: #15384 

CC: @oparoz @Xenopathic @MorrisJobke @PVince81 @DeepDiver1975

rullzer committed:
Move mimetypes.list.php to config/mimetypemapping.json

This allows users to add new mimetypemappings (extention -> mimetype)
themself. And not have to wait until a new release for updated
mimetypes.

Fixes: #15384

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/a44b0ffb-0c2b-4d0f-b686-c37ab73e76dd) was created.

rullzer commented:
mmmm, we actually should make sure that this file does not get overwritten during an upgrade... else the user loses his additions...
Same actually holds for config/mimetypealiases.json

oparoz commented:
Isn't the config folder left untouched?

Xenopathic commented:
Perhaps we should support loading multiple files, e.g. `mimetypemapping.custom.json`, which is packaged as empty and never gets overridden by updates. Any entries in that file override the ones in `mimetypemapping.json`, which has a note at the top explicitly telling admins not to put stuff there.

Or do it the other way around - make the custom one `mimetypemapping.json`, with a `mimetypemapping.dist.json` for the distributed packaged one.

Xenopathic commented:
But otherwise this PR looks good :+1:

oparoz commented:
Good call @Xenopathic. Makes it clear what is what.

Xenopathic commented:
Interesting side note, oC actually loads any config files with the name `*.config.php` (https://github.com/owncloud/core/blob/master/lib/private/config.php#L183). Not particularly useful for configs, but very handy in this case where we need a default packaged file that gets updated.

rullzer commented:
Agreed lets use mimetypemapping.dist.json... let me fix that for this PR. And create a new one for the mimetypealiases.json

rullzer committed:
Move mimetypemapping.json to mimetypemapping.dist.json

This allows users to create their own mapping file to extend our current
mappings. This makes sure that custom mappings are not lost on OC
upgrades.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/05e609f4-192c-4887-bde4-6a3a5e2129a1) was created.

rullzer commented:
Done :)

Xenopathic commented on config/mimetypemapping.dist.json:
```diff
@@ -0,0 +1,170 @@
+	"_comment2": "The first index in the mime type array is the assumed correct mimetype",
+	"_comment3": "and the second (if present] is a secure alternative",
+
+	"_comment4": "To add your own mappings create a file mimetypemapping.json",
```
This needs to be phrased like a warning:

> Any changes you make here will be overwritten on an update of ownCloud
> Put any custom mappings in a new file mimetypemapping.json in this directory

rullzer replied:
Done :)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/37959dc6-e35c-44e6-ae61-70b714a2f2fa) was created.

rullzer commented:
@owncloud-bot retest this please

MorrisJobke commented:
@owncloud-bot retest this please

Xenopathic commented:
Nice! :+1:

We need some documentation for this though. And do we want to ship a blank `mimetypemapping.json` with only comments saying how to use the file? Or do we want to require the user to manually create the file, and be expected to know the syntax etc?

oparoz commented:
@Xenopathic I'd say it's an advanced feature and it shouldn't be too much to ask of an admin to create the file.

rullzer commented:
Added todo issue for the docs.

One more review guys :)

MorrisJobke commented on lib/private/helper.php:
```diff
@@ -424,8 +424,18 @@ static function rmdirr($dir, $deleteSelf = true) {
 	 */
 	static public function getMimetypeDetector() {
 		if (!self::$mimetypeDetector) {
+			$dist = file_get_contents(OC::$SERVERROOT . '/config/mimetypemapping.dist.json');
```
@rullzer Can you change this to `\OC::$configDir`?

rullzer replied:
Ah didn't know about that one. Fixed!

rullzer committed:
Move to OC::$configDif

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2f9cfe74-ad0c-4da0-9bbd-9dd9c3e4eba1) was created.

rullzer commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/46a4e9b3-f991-4591-a64d-1ec01c87187b) was created.

Xenopathic commented on lib/private/helper.php:
```diff
@@ -424,8 +424,18 @@ static function rmdirr($dir, $deleteSelf = true) {
+
+			//Check if need to load custom mappings
+			if (file_exists(OC::$configDir . '/mimetypemapping.json')) {
+				$custom = file_get_contents(OC::$configDir . '/mimetypemapping.json');
```
Actually, TBH I think the code looks better with those slashes there. PHP ignores multiple slashes (as does any UNIX utility), so we have no bug concerns, and it also means less effort if `$configDir` at some point loses the trailing slash.

Xenopathic commented:
:+1:

rullzer commented:
@owncloud-bot retest this please

Xenopathic commented:
@owncloud-bot retest this please

MorrisJobke commented:
Mime type detection still works :+1:
